### ISSUE TITLE
NS_Node add_root is not thread safe

### ISSUE THREAD

**Issue by jokiefer** [django-treebeard/django-treebeard#278]
# problem
the current `NS_Node` model setup the necessary `tree_id` as simple [PositiveIntegerField](https://docs.djangoproject.com/en/4.2/ref/models/fields/#positiveintegerfield).

Inside the common tree functions such as `add_root` the lookup for the next free `tree_id` is a non thread safe implementation.
The lookup for the next free `tree_id` is based on the `get_last_root_node` function from the `Node` model which simply returns all root nodes without ordering (cause there is no [order_by](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#order-by) clause or [ordering](https://docs.djangoproject.com/en/4.2/ref/models/options/#ordering) configuration on models meta): 
https://github.com/django-treebeard/django-treebeard/blob/65eafcaa77cd0cfae2f68d199db4aa4c6a35a933/treebeard/ns_tree.py#L688

So the returned queryset will just return any root node. But that's not the kee problem why adding new root nodes is not thread safe.

If we dig deeper into the `add_root` method we can see how the next `free` id is calculated:

https://github.com/django-treebeard/django-treebeard/blob/65eafcaa77cd0cfae2f68d199db4aa4c6a35a933/treebeard/ns_tree.py#L153-L159

So if we adding new trees, by using the `add_root` for example, the next free id is just an increment of the last founded root_node.tree_id attribute.

If multiple trees are created in parallel running threads, this will result in corrupted tree data. 

# How to fix it?
My suggestion would be to add a [UniqueConstraint](https://docs.djangoproject.com/en/4.2/ref/models/constraints/#uniqueconstraint) to the `NS_Node`.
```python

class NS_Node(Node):
   class Meta:
      constraints = [
         models.UniqueConstraint(
            fields=["tree_id", "lft", "rgt"],
            name=%(app_label)s_%(class)s_unique_tree",
         )
      ]
```

If two threads creating a root node for a new tree with the same `tree_id`, `lft`, `rgt` values, this will result in a [IntegrityError](https://docs.djangoproject.com/en/4.1/ref/exceptions/#django.db.IntegrityError)

A fully thread safe version could be to extend the above code by refactoring the `NS_Node` model by using a [ForeignKey](https://docs.djangoproject.com/en/4.2/ref/models/fields/#foreignkey) to a primitive tree id table:

```python
class NS_TreeId(models.Model)
   # id would be the configured project autofield
   pass


```

AND refactoring the `add_root` method by using the transaction [atomic](https://docs.djangoproject.com/en/4.1/topics/db/transactions/#django.db.transaction.atomic) functionality like below:

```python
   @classmethod
    def add_root(cls, **kwargs):
        """Adds a root node to the tree."""

        # do we have a root node already?
        last_root = cls.get_last_root_node()

        if last_root and last_root.node_order_by:
            # there are root nodes and node_order_by has been set
            # delegate sorted insertion to add_sibling
            return last_root.add_sibling('sorted-sibling', **kwargs)

        with transaction.atomic():
           newtree_id = NS_TreeId.objects.create()

           if len(kwargs) == 1 and 'instance' in kwargs:
               # adding the passed (unsaved) instance to the tree
               newobj = kwargs['instance']
               if not newobj._state.adding:
                   raise NodeAlreadySaved("Attempted to add a tree node that is "\
                       "already in the database")
           else:
               # creating the new object
               newobj = get_result_class(cls)(**kwargs)

           newobj.depth = 1
           newobj.tree_id = newtree_id
           newobj.lft = 1
           newobj.rgt = 2
           # saving the instance before returning it
           newobj.save()
           return newobj
```

If it would be implemented like this, the database will always give us the next free id. Caused by the database transactional functionality, this would be tread safe. 

For now i do not determine all other side effects which the refactoring of the `tree _id` field would be causes.

I also think that other tree operating methods such as `add_sibling` are not thread safe cause the tree is not blocked by using the [select_for_update](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#select-for-update) methods.

**Comment by svenseeberg** [comment]
The (atomic) transaction feature is also needed for another reason: we're using treebeard a lot in https://github.com/digitalfabrik/integreat-cms and we frequently encounter inconsistent trees. Partially, the missing thread safety might be responsible. But there is also no guarantee, that a process does not die in the middle of rewriting a tree. So even without having multiple threads, a single transaction would be quite useful.

**Comment by timobrembeck** [comment]
Thanks for adding more context to this issue! However, this is probably a duplicate of #53, right?

**Comment by jokiefer** [comment]
#53 should be only a subset. Any tree operation should be done in one transaction and with blocking techniques such as select_for_update i guess.

**Comment by svenseeberg** [comment]
> ```python
>          models.UniqueConstraint(
>             fields=["tree_id", "lft", "rgt"],
>             name=%(app_label)s_%(class)s_unique_tree",
>          )
> ```

Would it not be better to use 2 separate unique constraints for (tree_id, lft) and (tree_id, rgt) ? In my experience there are cases in which only the lft or rgt value is wrong for a node and collides with another node.

**Comment by timobrembeck** [comment]
@jokiefer @svenseeberg Using `UniqueConstraint` is not going to work out of the box because this library uses raw sql queries which deliberately violate these constraints temporarily for performance reasons:

https://github.com/django-treebeard/django-treebeard/blob/f9674d0adf2ae3077c4475895f0df074179102c7/treebeard/ns_tree.py#L202-L209

This can result in e.g. SQL queries like

```
UPDATE "cms_page" SET tree_id = tree_id+1  WHERE tree_id >= 3
```

And as soon as the root node of `tree_id=3` is set to `4`, the unique constraint will complain that a root node with `tree_id=4` already exists. So even legitimate tree operations would cause error in this case.

There might be the (undocumented) method `connection.constraint_checks_disabled()` we can use, but I didn't try that out.

```
from django.db import connection
with connection.constraint_checks_disabled():
    do_stuff()
```

– https://stackoverflow.com/a/11926432

**Comment by svenseeberg** [comment]
As this would affect at least many operations, disabling the constraints temporarily basically would render them useless. The alternative would be to change the order of the operations so that the unique constraint would not be violated. But that would introduce even more code changes, I suppose.

**Comment by svenseeberg** [comment]
@timobrembeck found a possible solution: deferrable constraints. We created a draft implementation in https://github.com/digitalfabrik/integreat-cms/pull/2596/files

**Comment by solarissmoke** [comment]
Thanks everyone for the input on this issue. I have just opened #350 which wraps all of Treebeard's create/update operations in a transaction. That resolves part of this issue (exceptions caused by database constraints), but does not wholly prevent inconsistent state, and the documentation has been updated to suggest that projects implement additional locks/constraints depending on how critical consistency is for them, bearing in mind the performance implications of doing so.
